xapian-core/queryparser/mathtermgenerator_internal.h

```diff
@@ -0,0 +1,74 @@
+/** @file mathtermgenerator_internal.h
+ * @brief MathTermGenerator class internals
+ */
+/* Copyright (C) 2018 Guruprasad Hegde
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#ifndef XAPIAN_INCLUDED_MATHTERMGENERATOR_INTERNAL_H
+#define XAPIAN_INCLUDED_MATHTERMGENERATOR_INTERNAL_H
+
+#include "xapian/intrusive_ptr.h"
+#include <xapian/database.h>
+#include <xapian/document.h>
+#include <xapian/mathtermgenerator.h>
+
+#include <string>
+#include <vector>
+
+namespace Xapian {
+
+struct Symbol;
+
+class MathTermGenerator::Internal : public Xapian::Internal::intrusive_base {
+    friend class MathTermGenerator;
+    std::vector<Symbol> mrow;
+    void parse_mathml(const char *& ch);
+  public:
+    Internal() { }
+
+    void index_math(const char * ch);
+
+    // For debugging purpose.
+    std::size_t symbol_count() {
+	return mrow.size();
+    }
+};
+
+enum edge {
+    ADJACENT,
+    ABOVE,
+    BELOW,
+    OVER,
+    UNDER,
+    WITHIN,
+    ELEMENT
+};
+
+struct Symbol {
+    std::string label;
+    edge e;
+    // Symbols like fraction, matrix are represented over multiple lines.
+    // For example, consider a / b. a goes to top row, b goes to bottom
+    // row and fraction symbol is added in middle row.
+    // TODO Think of better way
+    std::vector<Symbol> trow;	// Top row.
+    std::vector<Symbol> brow;	// Bottom row.
+    explicit Symbol(std::string l, edge e_) : label(l), e(e_) { }
```

Comment:
`explicit` is only useful when the constructor has a single-argument form - this one doesn't.

Reply:
I believe that is no longer true in C++11. (See [this article](https://abseil.io/tips/142) for instance.) I think this means we should be making many more of our constructors explicit, so you can't do things like:

```c++
```

Reply:
Thanks, I wasn't aware of that.

The case for `explicit` for single argument constructors is rather stronger since pretty much any value passed or assigned could potentially be implicitly converted without it being obvious at the point it happens.  For the multi-argument case it needs to be a braced list it seems.

The "Constructors That Change Types, But Not Values [are OK as implicit]" idea which the article you link to suggests seems a sensible rule (or at least a rule of thumb).

Reply:
Thanks olly, james for pointing this out. I wasn't aware of that, forgot to remove `explicit` after changing constructor.

So we keep this as it is.